src/group_theory/quaternion_group.lean

```diff
@@ -0,0 +1,289 @@
+/-
+Copyright (c) 2021 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer (heavily based on `dihedral` by Shing Tak Lam)
+-/
+import data.zmod.basic
+import group_theory.order_of_element
+import data.nat.basic
+import tactic.interval_cases
+import group_theory.dihedral
+
+/-!
+# Quaternion Groups
+
+We define the (generalised) quaternion groups `quaternion_group n` of order `4n`, also known as
+dicyclic groups, with elements `a i` and `xa i` for `i : zmod n`.
+
+## Main definition
+
+`quaternion_group n`: The (generalised) quaternion group of order `4n`.
+
+## Implementation notes
+
+In mathematics, the name `quaternion_group` is reserved for the cases `n ≥ 2`. Since it would be
+inconvenient to carry around this condition we define `quaternion_group` also for `n = 0` and
+`n = 1`. `quaternion_group 0` is isomorphic to the infinite dihedral group, while
+`quaternion_group 1` is isomorphic to a cyclic group of order `4`.
+
+## References
+
+https://en.wikipedia.org/wiki/Dicyclic_group
+https://en.wikipedia.org/wiki/Quaternion_group
+
+## TODO
+
+Show that `quaternion_group 2 ≃* units (quaternion ℤ)`.
+
+-/
+
+/--
+The (generalised) quaternion group `quaternion_group n` of order `4n`. It can be defined by the
+presentation `⟨a, x | a²ⁿ = 1, x² = aⁿ, x⁻¹ax=a⁻¹⟩`. We write `a i` for `aⁱ` and `xa i` for `xaⁱ`.
+-/
+@[derive decidable_eq]
+inductive quaternion_group (n : ℕ) : Type
+| a : zmod (2 * n) → quaternion_group
+| xa : zmod (2 * n) → quaternion_group
+
+namespace quaternion_group
+
+variables {n : ℕ}
+
+/--
+Multiplication of the dihedral group.
+-/
+private def mul : quaternion_group n → quaternion_group n → quaternion_group n
+| (a i) (a j) := a (i + j)
+| (a i) (xa j) := xa (j - i)
+| (xa i) (a j) := xa (i + j)
+| (xa i) (xa j) := a (n + j - i)
+
+/--
+The identity `1` is given by `aⁱ`.
+-/
+private def one : quaternion_group n := a 0
+
+instance : inhabited (quaternion_group n) := ⟨one⟩
+
+/--
+The inverse of a an element of the dihedral group.
+-/
+private def inv : quaternion_group n → quaternion_group n
+| (a i) := a (-i)
+| (xa i) := xa (n + i)
+
+/--
+The group structure on `quaternion_group n`.
+-/
+instance : group (quaternion_group n) :=
+{ mul := mul,
+  mul_assoc :=
+  begin
+    rintros (i | i) (j | j) (k | k);
+    simp only [mul];
+    ring,
+    simp only [add_left_inj, add_right_inj],
```

Comment:
In some sense `ring` is overkill here, because you're just reasoning about addition. Unfortunately `abel` leaves a slightly messier state, and requires:
```suggestion
    abel,
    simp only [neg_mul_eq_neg_mul_symm, one_mul, int.cast_one, gsmul_eq_mul, int.cast_neg, add_right_inj],
```
I'll leave it up to you which one you prefer!